File: lesson_8/app/src/main/java/com/kalashnyk/denys/gmapsapp/di/module/ViewModelModule.java
package com.kalashnyk.denys.gmapsapp.di.module;

import android.app.Application;

import com.kalashnyk.denys.App;
import com.kalashnyk.denys.gmapsapp.di.scope.ViewModelScope;
import com.kalashnyk.denys.gmapsapp.domain.AllPinsViewModel;
import com.kalashnyk.denys.gmapsapp.domain.SinglePinViewModel;
import com.kalashnyk.denys.gmapsapp.repository.AppRepository;

import dagger.Module;
import dagger.Provides;

@Module
public class ViewModelModule {

    Application mApp;

    public ViewModelModule(App app) {
        mApp = app;
    }

    @ViewModelScope
    @Provides
    AllPinsViewModel providesAllPinsViewModel(AppRepository repository) {
        return new AllPinsViewModel(mApp, repository);
    }

    @ViewModelScope
    @Provides
    SinglePinViewModel providesSinglePinViewModel(AppRepository repository) {
        return new SinglePinViewModel(mApp, repository);
    }
}

File: lesson_9/app/src/main/java/com/kalashnyk/denys/kotlinbindingsample/di/module/ViewModelModule.kt
package com.kalashnyk.denys.kotlinbindingsample.di.module

import android.app.Application
import com.kalashnyk.denys.kotlinbindingsample.App
import com.kalashnyk.denys.kotlinbindingsample.di.scope.ViewModelScope
import com.kalashnyk.denys.kotlinbindingsample.domain.AllUsersViewModel
import com.kalashnyk.denys.kotlinbindingsample.domain.SingleUserViewModel
import com.kalashnyk.denys.kotlinbindingsample.repository.AppRepository
import dagger.Module
import dagger.Provides

@Module
class ViewModelModule(app: App) {

    internal var app: Application = app

    @ViewModelScope
    @Provides
    internal fun providesAllUserViewModel(repository: AppRepository): AllUsersViewModel {
        return AllUsersViewModel(app, repository)
    }

    @ViewModelScope
    @Provides
    internal fun providesSingleUserViewModel(repository: AppRepository): SingleUserViewModel {
        return SingleUserViewModel(app, repository)
    }
}